bouncer/src/repo/manipulator/modelconvertor/import/odaHelper/geometry_collector.cpp

```diff
@@ -177,23 +213,64 @@ repo::core::model::RepoNodeSet GeometryCollector::getMeshNodes() {
 
 repo::core::model::TransformationNode*  GeometryCollector::createTransNode(
 	const std::string &name,
-	const repo::lib::RepoUUID &parentId
-) {	
+	const repo::lib::RepoUUID &parentId) {	
 	return new repo::core::model::TransformationNode(repo::core::model::RepoBSONFactory::makeTransformationNode(repo::lib::RepoMatrix(), name, { parentId }));
 }
 
+bool GeometryCollector::hasMissingTextures()
+{
+	return missingTextures;
+}
 
+void GeometryCollector::getMaterialNodes(repo::core::model::RepoNodeSet& materials, repo::core::model::RepoNodeSet& textures) {
+	
+	materials.clear();
+	textures.clear();
 
-repo::core::model::RepoNodeSet GeometryCollector::getMaterialNodes() {
-	repo::core::model::RepoNodeSet matSet;
 	for (const auto &matPair : idxToMat) {
 		auto matIdx = matPair.first;
 		if (matToMeshes.find(matIdx) != matToMeshes.end()) {
-			matSet.insert(new repo::core::model::MaterialNode(matPair.second.cloneAndAddParent(matToMeshes[matIdx])));
+			auto& materialNode = matPair.second.first;
+			auto& textureNode = matPair.second.second;
+
+			auto matNode = new repo::core::model::MaterialNode(materialNode.cloneAndAddParent(matToMeshes[matIdx]));
+			materials.insert(matNode);
+
+			if (!textureNode.isEmpty()) {
+				auto texNode = new repo::core::model::TextureNode(textureNode.cloneAndAddParent(matNode->getSharedID()));
```

Comment:
You already know what's the material SharedID is when you created the Texture node (You create the material node before the texture node). So you should create the node with right parent in the first place, avoiding another cloning.